libs/core/algorithms/include/hpx/parallel/unseq/simd_helpers.hpp

```diff
@@ -0,0 +1,165 @@
+//  Copyright (c) 2022 A Kishore Kumar
+//  Copyright (c) 2023 Hartmut Kaiser
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/functional/detail/invoke.hpp>
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+#include <hpx/parallel/unseq/reduce_helpers.hpp>
+#include <hpx/type_support/construct_at.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+namespace hpx::parallel::util {
+    /*
+        Compiler and Hardware should also support vector operations for IterDiff,
+        else we see slower performance when compared to sequential version
+    */
+    template <class Iter, class IterDiff, class F>
+    Iter unseq_first_n(Iter const first, IterDiff const n, F&& f_) noexcept
+    {
+        /*
+            OMP loops can not have ++Iter, only integral types are allowed
+            Hence perform arthemetic on Iterators
+            which is O(1) only in case of random access iterators
+        */
+        static_assert(hpx::traits::is_random_access_iterator_v<Iter>,
+            "algorithm is efficient only in case of Random Access Iterator");
+#if HPX_EARLYEXIT_PRESENT
+        IterDiff i = 0;
+        // clang-format off
+        HPX_PRAGMA_VECTOR_UNALIGNED HPX_PRAGMA_SIMD_EARLYEXIT
+        for (; i < n; ++i)
+        {
+            if (f(first, i))
+            {
+                break;
+            }
+        }
+        // clang-format on
+
+        return first + i;
+#else
+        auto f = [f_ = HPX_FORWARD(F, f_)](
+                     auto it, int ind) { return f_(*(ind + it)); };
+        // std::int32_t has best support for vectorization from compilers and hardware
+        IterDiff i = 0;
+        static constexpr std::int32_t num_blocks =
+            HPX_LANE_SIZE / sizeof(std::int32_t);
+        alignas(HPX_LANE_SIZE) std::int32_t simd_lane[num_blocks] = {0};
+        while (i <= n - num_blocks)
+        {
+            std::int32_t found_flag = 0;
+
+            // clang-format off
+            HPX_PRAGMA_VECTOR_UNALIGNED HPX_VECTOR_REDUCTION(| : found_flag)
+            for (IterDiff j = i; j < i + num_blocks; ++j)
+            {
+                std::int32_t const t = f(first, j);
```

Comment:
Can we remove the lambda altogether?
```suggestion
                std::int32_t const t = f_(*(first + j));
```

Reply:
Requested change has been made